### PR TITLE
improve call performance

### PR DESCRIPTION
**Pull request by Hades32**
sorry, for this rather random assortment of changes. This mostly stemmed from me wondering why we need to decode the JSON that we are going to send out again...

Maybe you'll only want to pick some changes as you see fit.

Also sorry for changing formatting in the same commit, but I didn't expect that you hadn't setup `go fmt` for everything.

* fixed leakage of never closed http.Client requests
* remove unnecessary string/byte casts
* remove unnecessary JSON decoding (including reflection)
* remove unnecessary http handler indirection & very implicit use of DefaultMux
* use idiomatic http error functions
* get rid of unidiomatic utils package

### CODE REVIEW

File: handler/utils.go [hunk 1]
```diff
@@ -2,19 +2,21 @@ package handler
...
+type ResponseHTTP struct {
+	StatusCode int               `json:"statusCode"`
+	Body       json.RawMessage   `json:"body"`
+	Headers    map[string]string `json:"headers"`
 }
 
 // GetResponse - Transform a response string into an HTTP Response structure
-func GetResponse(response string) (handlerResponse *ResponseHTTP, err error) {
-	if err := json.Unmarshal([]byte(response), &handlerResponse); err != nil {
+func GetResponse(response io.Reader) (*ResponseHTTP, error) {
```
Comment (AmFlint) [inline handler/utils.go:17]: Why did you remove the named return parameters ?

File: server/server.go [hunk 2]
```diff
@@ -66,29 +64,24 @@ func Start() error {
...
-	}
-
 	requestHandler, err := buildRequestHandler()
 	if err != nil {
 		// TODO: FORMAT ERROR
 		return err
 	}
 
-	http.HandleFunc("/", requestHandler)
+	s := &http.Server{
```
Comment (AmFlint) [inline server/server.go:73]: You talked about using `http.NewServer` in your other PullRequest, can you use this method in this PR please ?

File: server/server.go [hunk 3]
```diff
@@ -111,42 +104,43 @@ func buildRequestHandler() (func(http.ResponseWriter, *http.Request), error) {
 		// Authenticate function, if an error occurs, do not execute the handler
 		if err := authentication.Authenticate(request); err != nil {
 			log.Print(err)
-			writeResponse(response, http.StatusNotFound, "")
+			http.NotFound(response, request) // NOTE: why not found? should be a proper middleware and handle 401/403
```
Comment (AmFlint) [inline server/server.go:107]: We're using Not Found here because the function should not be executed, and we do not want to expose users's private functions to external users.

For example, if you try to send a request to test-function-does-not-exist.functions.fnc.fr-par.scw.cloud, you'll get 404 because this function does not exist on our Platform. 

If you try to access a private function without JWT, we do not want to tell you This functions EXISTS but you do not have the permissions to execute it.

Do you think this is not the right way to do it ?

File: server/server.go [hunk 4]
```diff
@@ -111,42 +104,43 @@ func buildRequestHandler() (func(http.ResponseWriter, *http.Request), error) {
...
-			writeResponse(response, http.StatusBadRequest, err.Error())
+			http.Error(response, err.Error(), http.StatusBadRequest)
 			return
 		}
 
 		// 3: Format event and context
 		event, err := events.FormatEvent(request, triggerType)
 		if err != nil {
-			writeResponse(response, http.StatusInternalServerError, err.Error())
+			http.Error(response, err.Error(), http.StatusInternalServerError)
```
Comment (AmFlint) [inline server/server.go:121]: I did not know this method existed haha Good to know

File: server/server.go [hunk 5]
```diff
@@ -156,13 +150,15 @@ func buildRequestHandler() (func(http.ResponseWriter, *http.Request), error) {
...
 
-		// Manage the case where user returns either a string (or stringified JSON) response body, or any other type
-		responseBody, err := utils.GetStringFromInterface(handlerRes.Body)
-		if err != nil {
-			writeResponse(response, http.StatusInternalServerError, errorInvalidResponseBody.Error())
-			return
+		response.WriteHeader(handlerRes.StatusCode)
+		// when lambda returns a string as body it expects to return it without json encoding
+		// NOTE: this should be improved by setting a proper content-type in the lambda
+		if handlerRes.Body[0] == '"' {
```
Comment (AmFlint) [inline server/server.go:156]: I'm not sure this is the best way to check if handler's response is of type String

File: handler/utils.go (hunk 1 above)
Comment (Hades32) [reply handler/utils.go:17]: the method was using mixed return styles, which I try to avoid has it makes it even harder to understand the code for newcomers. (I generally use named returns sparingly, but that's just personal preference)

File: server/server.go (hunk 2 above)
Comment (Hades32) [reply server/server.go:73]: that was `http.NewRequest`. there is no such method for the server

File: server/server.go (hunk 3 above)
Comment (Hades32) [reply server/server.go:107]: hmm, I can understand the argumentation, but I think it will be VERY confusing to a user who (due some bug on his side) sends you an old, expired JWT and will return 404 NOT FOUND... I once debugged a similar system and it took me far too long to understand the error.

I would argue that this approach is _security by obscurity_. Either your authenticate is safe and knowing the endpoint doesn't matter or it is not, and hiding it this way will only help so much, because probably the URL is visible in some JavaScript file of the using WebSite anyway.

And my comment about the middleware:
The code in `authentication` talks about being a middleware, but it's not actually implemented as a http middleware (i.e. a method taking a `http.Handler` and also returning one). But as it is only used in exactly one place this probably doesn't matter too much.

File: server/server.go (hunk 5 above)
Comment (Hades32) [reply server/server.go:156]: yes, I also still feel unsure about that one. But I didn't find a different way to check and at the same time prevent the unnecessary parsing.

I wrote a small test and the JSON lib seems guarantee returning no extra whitespace or other things around the string. So it _should_ work...

Maybe you can do a perf test in your environment to see if it's worth it

File: server/server.go (hunk 4 above)
Comment (Hades32) [reply server/server.go:121]: learned about it pretty recently, too ;)

File: handler/utils.go (hunk 1 above)
Comment (Hades32) [reply handler/utils.go:17]: BTW, was this API design chosen to be compatible with AWS? Because it generally seems pretty inefficient. Just defining custom HTTP response headers and always passing the body through would have been much simpler IMHO

Comment (AmFlint) [reply handler/utils.go:17]: Yes, we wanted to make it easy for users to migrate from AWS Lambda.

Also, we want to take advantage of the knowledge serverless users acquired by using AWS Lambda, so they can use it on this product.

We could probably update our sub-runtimes to do this logic, so core-runtime might pass the body through, but it will be a pain to maintain, as instead of coding it once in the core runtime, we'll have to code (and maintain) it in every sub runtimes... Which is why I decided to do it in core runtime and sacrifice performance

File: server/server.go (hunk 2 above)
Comment (AmFlint) [reply server/server.go:73]: Oh, I see

File: server/server.go (hunk 3 above)
Comment (AmFlint) [reply server/server.go:107]: Yes, it works like a middleware, but is not implemented like an HTTP Middleware per se.

I understand your point and I believe it is the best thing to do: Can you update the error code to notAuthorized please ?

File: server/server.go (hunk 5 above)
Comment (AmFlint) [reply server/server.go:156]: We will have to test this with different return types:
- Handler returns { body: 'string' }
- handler returns { body: '{"test": "test"}'
- handler returns { body: JSON.stringify({test: 'test'}) }
- handler returns { body: true }
- handler returns { body: 1 }

Comment (Hades32) [reply server/server.go]: add tests for all given examples and they work just fine

File: server/server.go (hunk 3 above)
Comment (Hades32) [reply server/server.go]: I changed it to return 401/403 depending on error case
